File: backend/pkg/database/repository/source.go
package repository

import (
	"context"
	"encoding/json"
	"fmt"

	"backend/pkg/core"
	"backend/pkg/core/source"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"github.com/google/uuid"
	"github.com/pgvector/pgvector-go"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// SourceRepository provides data access for article sources
type SourceRepository struct {
	db *gorm.DB
}

// NewSourceRepository creates a new SourceRepository
func NewSourceRepository(db *gorm.DB) *SourceRepository {
	return &SourceRepository{db: db}
}

// sourceModelToType converts a database model to types
func sourceModelToType(m *models.Source) *types.Source {
	var metaData map[string]interface{}
	if m.MetaData != nil {
		_ = json.Unmarshal(m.MetaData, &metaData)
	}

	var embedding []float32
	if m.Embedding.Slice() != nil {
		embedding = m.Embedding.Slice()
	}

	return &types.Source{
		ID:         m.ID,
		ArticleID:  m.ArticleID,
		Title:      m.Title,
		Content:    m.Content,
		URL:        m.URL,
		SourceType: m.SourceType,
		Embedding:  embedding,
		MetaData:   metaData,
		CreatedAt:  m.CreatedAt,
	}
}

// sourceTypeToModel converts a types type to database model
func sourceTypeToModel(s *types.Source) *models.Source {
	var metaData datatypes.JSON
	if s.MetaData != nil {
		data, _ := json.Marshal(s.MetaData)
		metaData = datatypes.JSON(data)
	}

	var embedding pgvector.Vector
	if len(s.Embedding) > 0 {
		embedding = pgvector.NewVector(s.Embedding)
	}

	return &models.Source{
		ID:         s.ID,
		ArticleID:  s.ArticleID,
		Title:      s.Title,
		Content:    s.Content,
		URL:        s.URL,
		SourceType: s.SourceType,
		Embedding:  embedding,
		MetaData:   metaData,
		CreatedAt:  s.CreatedAt,
	}
}

// FindByID retrieves a source by its ID
func (r *SourceRepository) FindByID(ctx context.Context, id uuid.UUID) (*types.Source, error) {
	var model models.Source
	if err := r.db.WithContext(ctx).First(&model, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return sourceModelToType(&model), nil
}

// FindByArticleID retrieves all sources for an article
func (r *SourceRepository) FindByArticleID(ctx context.Context, articleID uuid.UUID) ([]types.Source, error) {
	var sourceModels []models.Source
	if err := r.db.WithContext(ctx).Where("article_id = ?", articleID).Order("created_at DESC").Find(&sourceModels).Error; err != nil {
		return nil, err
	}

	sources := make([]types.Source, len(sourceModels))
	for i, m := range sourceModels {
		sources[i] = *sourceModelToType(&m)
	}
	return sources, nil
}

// sourceWithArticleRow represents the joined query result
type sourceWithArticleRow struct {
	models.Source
	ArticleTitle string `gorm:"column:article_title"`
	ArticleSlug  string `gorm:"column:article_slug"`
}

// List retrieves all sources with pagination and article metadata
func (r *SourceRepository) List(ctx context.Context, opts source.SourceListOptions) ([]source.SourceWithArticle, int64, error) {
	if opts.Page < 1 {
		opts.Page = 1
	}
	if opts.PerPage < 1 || opts.PerPage > 100 {
		opts.PerPage = 20
	}

	offset := (opts.Page - 1) * opts.PerPage

	var total int64
	if err := r.db.WithContext(ctx).Model(&models.Source{}).Count(&total).Error; err != nil {
		return nil, 0, err
	}

	var rows []sourceWithArticleRow
	err := r.db.WithContext(ctx).Table("article_source").
		Select("article_source.*, article.draft_title as article_title, article.slug as article_slug").
		Joins("LEFT JOIN article ON article.id = article_source.article_id").
		Order("article_source.created_at DESC").
		Offset(offset).
		Limit(opts.PerPage).
		Scan(&rows).Error

	if err != nil {
		return nil, 0, err
	}

	results := make([]source.SourceWithArticle, len(rows))
	for i, row := range rows {
		results[i] = source.SourceWithArticle{
			Source:       *sourceModelToType(&row.Source),
			ArticleTitle: row.ArticleTitle,
			ArticleSlug:  row.ArticleSlug,
		}
	}

	return results, total, nil
}

// SearchSimilar performs vector similarity search for sources within an article
func (r *SourceRepository) SearchSimilar(ctx context.Context, articleID uuid.UUID, embedding []float32, limit int) ([]types.Source, error) {
	var sourceModels []models.Source

	embeddingVector := pgvector.NewVector(embedding)
	query := fmt.Sprintf(
		"SELECT * FROM article_source WHERE article_id = '%s' AND embedding IS NOT NULL ORDER BY embedding <-> '%s' LIMIT %d",
		articleID.String(),
		embeddingVector.String(),
		limit,
	)

	if err := r.db.WithContext(ctx).Raw(query).Scan(&sourceModels).Error; err != nil {
		return nil, err
	}

	sources := make([]types.Source, len(sourceModels))
	for i, m := range sourceModels {
		sources[i] = *sourceModelToType(&m)
	}
	return sources, nil
}

// Save creates a new source
func (r *SourceRepository) Save(ctx context.Context, source *types.Source) error {
	model := sourceTypeToModel(source)
	if model.ID == uuid.Nil {
		model.ID = uuid.New()
		source.ID = model.ID
	}
	return r.db.WithContext(ctx).Create(model).Error
}

// Update updates an existing source
func (r *SourceRepository) Update(ctx context.Context, source *types.Source) error {
	model := sourceTypeToModel(source)
	return r.db.WithContext(ctx).Save(model).Error
}

// Delete removes a source by its ID
func (r *SourceRepository) Delete(ctx context.Context, id uuid.UUID) error {
	result := r.db.WithContext(ctx).Delete(&models.Source{}, id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return core.ErrNotFound
	}
	return nil
}

File: backend/src/database/repository/source.rs
use async_trait::async_trait;
use diesel::{ExpressionMethods, OptionalExtension, QueryDsl, SelectableHelper, dsl::count_star};
use diesel_async::RunQueryDsl;
use pgvector::{Vector, VectorExpressionMethods};
use serde_json::Value;
use uuid::Uuid;

use crate::{
    core::{
        article::ArticleRepository,
        source::{
            ArticleLookupPort, MetaData, Source, SourceListOptions, SourceRepository,
            SourceWithArticle,
        },
    },
    database::{
        models::source::{NewSourceRow, SourceChangeset, SourceRow},
        pool::PgPool,
        repository::article::DieselArticleRepository,
    },
    error::AppError,
    schema::{article, article_source},
};

#[derive(Clone)]
pub struct DieselSourceRepository {
    pool: PgPool,
}

#[async_trait]
impl ArticleLookupPort for DieselArticleRepository {
    async fn ensure_exists(&self, article_id: Uuid) -> Result<(), AppError> {
        ArticleRepository::find_by_id(self, article_id)
            .await
            .map(|_| ())
    }
}

impl DieselSourceRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }
}

#[async_trait]
impl SourceRepository for DieselSourceRepository {
    async fn find_by_id(&self, id: Uuid) -> Result<Source, AppError> {
        let mut connection = self.connection().await?;
        article_source::table
            .find(id)
            .select(SourceRow::as_select())
            .first(&mut connection)
            .await
            .optional()
            .map_err(|_| AppError::Database)?
            .ok_or(AppError::NotFound)
            .map(Into::into)
    }

    async fn find_by_article_id(&self, article_id: Uuid) -> Result<Vec<Source>, AppError> {
        let mut connection = self.connection().await?;
        article_source::table
            .filter(article_source::article_id.eq(article_id))
            .order(article_source::created_at.desc())
            .select(SourceRow::as_select())
            .load(&mut connection)
            .await
            .map(|rows: Vec<SourceRow>| rows.into_iter().map(Into::into).collect())
            .map_err(|_| AppError::Database)
    }

    async fn list(
        &self,
        options: SourceListOptions,
    ) -> Result<(Vec<SourceWithArticle>, i64), AppError> {
        let page = options.page.max(1);
        let per_page = if !(1..=100).contains(&options.per_page) {
            20
        } else {
            options.per_page
        };
        let mut connection = self.connection().await?;
        let total = article_source::table
            .select(count_star())
            .first(&mut connection)
            .await
            .map_err(|_| AppError::Database)?;
        let rows = article_source::table
            .order(article_source::created_at.desc())
            .offset((page - 1) * per_page)
            .limit(per_page)
            .select(SourceRow::as_select())
            .load::<SourceRow>(&mut connection)
            .await
            .map_err(|_| AppError::Database)?;
        let mut result = Vec::with_capacity(rows.len());
        for row in rows {
            let article_data = article::table
                .find(row.article_id)
                .select((article::draft_title, article::slug))
                .first::<(Option<String>, String)>(&mut connection)
                .await
                .optional()
                .map_err(|_| AppError::Database)?;
            let (title, slug) = article_data.unwrap_or_default();
            result.push(SourceWithArticle {
                source: row.into(),
                article_title: title.unwrap_or_default(),
                article_slug: slug,
            });
        }
        Ok((result, total))
    }

    async fn save(&self, source: &mut Source) -> Result<(), AppError> {
        if source.id.is_nil() {
            source.id = Uuid::new_v4();
        }
        let mut connection = self.connection().await?;
        diesel::insert_into(article_source::table)
            .values(new_row(source))
            .execute(&mut connection)
            .await
            .map(|_| ())
            .map_err(|_| AppError::Database)
    }

    async fn update(&self, source: &Source) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        let affected = diesel::update(article_source::table.find(source.id))
            .set(changeset(source))
            .execute(&mut connection)
            .await
            .map_err(|_| AppError::Database)?;
        if affected == 0 {
            diesel::insert_into(article_source::table)
                .values(new_row(source))
                .execute(&mut connection)
                .await
                .map_err(|_| AppError::Database)?;
        }
        Ok(())
    }

    async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        match diesel::delete(article_source::table.find(id))
            .execute(&mut connection)
            .await
            .map_err(|_| AppError::Database)?
        {
            0 => Err(AppError::NotFound),
            _ => Ok(()),
        }
    }

    async fn search_similar(
        &self,
        article_id: Uuid,
        embedding: &[f32],
        limit: i64,
    ) -> Result<Vec<Source>, AppError> {
        let mut connection = self.connection().await?;
        let mut query = article_source::table
            .filter(article_source::article_id.eq(article_id))
            .filter(article_source::embedding.is_not_null())
            .order(article_source::embedding.l2_distance(Vector::from(embedding.to_vec())))
            .into_boxed();
        if limit >= 0 {
            query = query.limit(limit);
        }
        query
            .select(SourceRow::as_select())
            .load(&mut connection)
            .await
            .map(|rows: Vec<SourceRow>| rows.into_iter().map(Into::into).collect())
            .map_err(|_| AppError::Database)
    }
}

impl From<SourceRow> for Source {
    fn from(row: SourceRow) -> Self {
        Self {
            id: row.id,
            article_id: row.article_id,
            title: row.title.unwrap_or_default(),
            content: row.content,
            url: row.url.unwrap_or_default(),
            source_type: row.source_type.unwrap_or_default(),
            embedding: row.embedding.map(|value| value.to_vec()),
            meta_data: row.meta_data.and_then(metadata),
            created_at: row.created_at,
        }
    }
}

fn new_row(source: &Source) -> NewSourceRow {
    NewSourceRow {
        id: source.id,
        article_id: source.article_id,
        title: Some(source.title.clone()),
        content: source.content.clone(),
        url: Some(source.url.clone()),
        source_type: Some(source.source_type.clone()),
        embedding: source.embedding.clone().map(Vector::from),
        meta_data: metadata_value(source.meta_data.as_ref()),
        created_at: source.created_at.unwrap_or_else(chrono::Utc::now),
    }
}

fn changeset(source: &Source) -> SourceChangeset {
    SourceChangeset {
        article_id: source.article_id,
        title: Some(source.title.clone()),
        content: source.content.clone(),
        url: Some(source.url.clone()),
        source_type: Some(source.source_type.clone()),
        embedding: source.embedding.clone().map(Vector::from),
        meta_data: source
            .meta_data
            .as_ref()
            .map(|value| metadata_value(Some(value))),
    }
}

fn metadata(value: Value) -> Option<MetaData> {
    match value {
        Value::Object(map) => Some(map.into_iter().collect()),
        _ => None,
    }
}

fn metadata_value(value: Option<&MetaData>) -> Value {
    Value::Object(
        value
            .map(|map| map.clone().into_iter().collect())
            .unwrap_or_default(),
    )
}
